### PR TITLE
build: upload on bintray from travis when a git tag is set

### PR DESCRIPTION
**Pull request by davidB**
The commit will allow to publish (in dry mode) from travis to bintray in repository https://bintray.com/jmonkeyengine/org.jmonkeyengine when a tag (starting with jmeMainVersion) is pushed.

## Setup
* On bintray
  * create account on bintray (eg using github identity provider )
  * ask me or Erlend to register the account as member of the organization jmonkeyengine to be able to upload artifacts
* On travis
  * register 2 env variables (I prefer to setup in the settings'panel, but I guess using secure in .travis.yml should work)
    * `ORG_GRADLE_PROJECT_bintray_user` with the bintray account name, could be displayed in build.log
    * `ORG_GRADLE_PROJECT_bintray_api_key` with the bintray api key of the account (from https://bintray.com/profile/edit), could NOT be displayed in build.log 

## Apply:
* On desktop
set a git tag that start by jmeMainVersion (3.1 today) and push.
```
git tag 3.1.0-alpha2
git push --tags
```
The tag will be used as version and jmePomVersion
* After build connect + login on bintray, the artifact should be uploaded in the repostory but not published (I disable the auto-publish (dry = false) until we test the process). So you can choose to publish or to discard the uploaded artifacts. Login is required to see the panel or info about "uploaded but not published data)

## Try on local:

* setup the env variable or setup bintray_user and bintray_api_key into your $HOME/.gradle/gradle.properties
* git tag or define TRAVIS_TAG (env variable) with a value that start by "3.1" (eg "3.1.0-myversion")
* call `./gradlew bintrayUpload`

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -17,6 +18,9 @@ apply from: file('upload.gradle')
 subprojects {
     if(!project.name.equals('jme3-android-examples')) {
         apply from: rootProject.file('common.gradle')
+        if (!['jme3-testdata', 'sdk'].contains(project.name)) {
```
Comment (kvainer) [inline build.gradle]: What about jbullet / nifty? We shouldn't upload those because they reference external artifacts.

Comment (shadowislord) [inline build.gradle]: What about jbullet / nifty? We shouldn't upload those because they reference external artifacts.

File: bintray.gradle [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+//
+
+apply plugin: 'com.jfrog.bintray'
+
+bintray {
+    user = bintray_user
+    key = bintray_api_key
+    configurations = ['archives']
+    pkg {
+        repo = 'org.jmonkeyengine'
```
Comment (shadowislord) [inline bintray.gradle]: Why not use the POM_ constants for everything? We want the POMs in our repo and the internal repo to be as similar as possible

File: build.gradle (hunk 1 above)
Comment (davidB) [inline build.gradle]: we could upload them, so there are available on the org.jmonkeyengine repository. But choose to not "link" them to jcenter. linking is a per artifact action, that should be done on bintray.

File: bintray.gradle (hunk 2 above)
Comment (davidB) [inline bintray.gradle]: I don't understand. the 'repo' is not the groupId, 'repo' + 'userOrg' define the destination repository on bintray.

Comment (davidB) [inline bintray.gradle]: bintray is not limited to host maven repository / artifact.

Comment (shadowislord) [inline bintray.gradle]: By repo I meant jcenter and updates.jmonkeyengine.com/maven

Comment (davidB) [inline bintray.gradle]: updates.jmonkeyengine.com/maven is not a bintray repo
upload directly to jcenter is not the recommended path, we upload into our own repository, and then linked artifacts are availble in every linked repository like jcenter. eg I upload my artifact in my own repository and for artifacts linked into jmonkeyengine/contrib repository or jcenter.
see [Exposing Your Package to the World through More Popular Repositories (Linking a Package)](https://bintray.com/docs/usermanual/uploads/uploads_exposingyourpackagetotheworldthroughmorepopularrepositorieslinkingapackage.html)

**Comment by davidB** [comment]
I didn't check the generated pom. There are missing info to be maven central compliant. I'll complete them.

File: .travis.yml [hunk 3]
```diff
@@ -54,3 +54,6 @@ before_install:
   # wget http://dl.google.com/android/ndk/android-ndk-r10c-linux-x86_64.bin -O ndk.bin
   # 7z x ndk.bin -y > /dev/null
   # export ANDROID_NDK=`pwd`/android-ndk-r10c
+
+after_success:
+  - '[ -n "$TRAVIS_TAG"" ] && [ "$TRAVIS_PULL_REQUEST" == "false" ] && ./gradlew bintrayUpload || :'
```
Comment (kvainer) [inline .travis.yml]: Why not put this in the `script: ` section? Alternatively, why not put both upload lines in the `after_success` section?

Comment (shadowislord) [inline .travis.yml]: Why not put this in the `script: ` section? Alternatively, why not put both upload lines in the `after_success` section?

File: common.gradle [hunk 4]
```diff
@@ -61,12 +61,85 @@ task javadocJar(type: Jar, dependsOn: javadoc, description: 'Creates a jar from
...
+    }
+    licenses {
+        license {
+            name POM_LICENSE_NAME
+            url POM_LICENSE_URL
+            distribution POM_LICENSE_DISTRIBUTION
+        }
+    }
+    // from http://hub.jmonkeyengine.org/introduction/team/
+    developers {
```
Comment (shadowislord) [inline common.gradle]: Is this `developres` section actually required? OSS projects have many contributors and we can't give credit to everybody.

File: .travis.yml (hunk 3 above)
Comment (davidB) [inline .travis.yml]: ok to push the 2 upload in the after_succes section (seems to be the convention from sample I saw)

Comment (shadowislord) [inline .travis.yml]: sounds good

File: common.gradle (hunk 4 above)
Comment (davidB) [inline common.gradle]: Yes it is required to push to maven central.

Comment (shadowislord) [inline common.gradle]: Just add a name 'jMonkeyEngine Team' with id 'jMonkeyEngine' then?

File: version.gradle [hunk 5]
```diff
@@ -76,7 +76,11 @@ task configureVersionInfo {
         jmeFullVersion = jmeMainVersion
         jmePomVersion  = jmeVersion
         
-        if (jmeBranchName != "master") {
+        if (jmeGitTag.startsWith(jmeMainVersion)) {
```
Comment (shadowislord) [inline version.gradle]: Can you update the 4 examples in the top comment to match this new behavior? I want to see how it looks like now

File: .travis.yml [hunk 6]
```diff
@@ -54,3 +53,7 @@ before_install:
   # wget http://dl.google.com/android/ndk/android-ndk-r10c-linux-x86_64.bin -O ndk.bin
   # 7z x ndk.bin -y > /dev/null
   # export ANDROID_NDK=`pwd`/android-ndk-r10c
+
+after_success:
+  - '[ "$TRAVIS_BRANCH" == "master" ] && [ "$TRAVIS_PULL_REQUEST" == "false" ] && ./gradlew uploadArchives || :'
+  - '[ -n "$TRAVIS_TAG"" ] && [ "$TRAVIS_PULL_REQUEST" == "false" ] && ./gradlew bintrayUpload || :'
```
Comment (shadowislord) [inline .travis.yml]: There two quotes here: 
`"$TRAVIS_TAG""`

File: build.gradle (hunk 1 above)
Comment (shadowislord) [inline build.gradle]: OK, that works

**Pull request merged**